### PR TITLE
Better python binary detection

### LINKED ISSUE #2090: compiling syslog-ng with python3
# syslog-ng

## Version of syslog-ng

syslog-ng git head from today

## Platform

openSUSE Leap 15.0

# Issue

```
[    5s] checking for python... no
[    5s] configure: error: Python interpreter is required
[    5s] error: Bad exit status from /var/tmp/rpm-tmp.9HbhpI (%build)
```
Configure fails, if **python**  binary is not available. That is the case on openSUSE Leap 15.0 and SLES, where only python3 is installed by default. The **python** binary is provided by the Python 2.7 package. If I symlink python to python3, compilation seems to work.
I have tried to use --with-python=python3 and --with-python=3.6, but neither of these resolved the problem.

### PR DESCRIPTION
**Pull request by algernon**
Instead of using `AC_PATH_PROG` to find the python binary (which can only find binaries named `python`, not `python3` and similar), use `AM_PATH_PYTHON` instead, and take `--with-python` into account as well.

For this to work, we also need to call a few of our python scripts through the interpreter, to avoid having to generate the scripts.

Fixes #2090.

### CODE REVIEW

**Comment by czanik** [comment]
Thanks! I am on mobile net right now but will check package building once back in my office.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by czanik** [comment]
Thanks! **configure** ran fine and package is created for openSUSE Leap 15.0.

Next step (only tomorrow): testing the results.

File: Mk/lex-rules.am [hunk 1]
```diff
@@ -1,6 +1,6 @@
 %.y: %.ym $(syslog_ng_tools)/merge-grammar.py $(syslog_ng_tools)/cfg-grammar.y
 	$(AM_V_at) $(mkinstalldirs) $(dir $@)
-	$(AM_V_GEN) $(syslog_ng_tools)/merge-grammar.py $< > $@
+	$(AM_V_GEN) $(PYTHON) $(syslog_ng_tools)/merge-grammar.py $< > $@
```
Comment (Kokan) [inline Mk/lex-rules.am]: The CMake also has to use PYTHO for the grammar merge. Hint: `cmake/Modules/GenerateYFromYm.cmake`.

Comment (algernon) [reply Mk/lex-rules.am]: Thanks, I missed that. Sorry!

Comment (algernon) [reply Mk/lex-rules.am]: Fixed (I hope).

Comment (Kokan) [reply Mk/lex-rules.am]: Maybe my comment was misleading the PYTHON variable in cmake does not exists, and does not point to the python interpreter. The `PYTHON_EXECUTABLE` does.

But still there is also a reason why cmake version won't work as intended.

The `with-python` and the option in cmake was not intended to specify which python version you are using for build, but rather for the module.

In cmake the python detection was "hidden" from the upper level, and determined in the python module. To follow up the same change in cmake as in autotools, one probably should move the python detection to the upper level in cmake.

Comment (Kokan) [reply Mk/lex-rules.am]: Sure, I hope on next Monday I am going to be in the office, and we can check it together.

**Comment by bazsi** [comment]
@kira-syslogng test this please

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by algernon** [comment]
Updated. Moved the CMake python detection out into the top-level `CMakeLists.txt`, so that `PYTHON_EXECUTABLE` will be available for the grammar generation stuff too.

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: modules/python/CMakeLists.txt [hunk 2]
```diff
@@ -1,9 +1,4 @@
-set (PYTHON_VERSION "AUTO" CACHE STRING "Version of the installed development library" )
-
 if ("${PYTHON_VERSION}" STREQUAL "AUTO")
```
Comment (Kokan) [inline modules/python/CMakeLists.txt]: This condition is not needed anymore.

**Comment by algernon** [comment]
Oof. I'll see about that, but that should be a separate PR, in my opinion. It "only" affects the tests, right?

**Comment by Kokan** [comment]
No, it is all of the python module, well it says something like that, if the `PYTHON_VERSION` is auto, to whatever (do not check for python, but tries to compile).

**Comment by algernon** [comment]
> No, it is all of the python module, well it says something like that, if the PYTHON_VERSION is auto, to whatever (do not check for python, but tries to compile).

Should have been clearer, my comment was a reply to [@mochrul's](https://github.com/balabit/syslog-ng/pull/2092#issuecomment-399756282).

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Pull request merged**